### DISCUSSION TITLE
odd behaviour with a kettle....

### DISCUSSION THREAD

**Discussion by takigama** [jasonacox/tinytuya#131] [General]
I have a tuya based kettle that i've been trying to figure out how to make work. It was originally only displaying one DP:

```
paulr@graviton:~/tmp/tuya$ python test.py
set_status() result {u'devId': u'xxxxxxx', u'dps': {u'5': 94}}
detect_available_dps() result {u'5': 94}
```
The dp (5) is actually the temperature of the water (id 5), i tried sending random data to it on dp 5, but that didnt really do much (it does have a "hold this temperature" control so I thought maybe it might be a way to tell it to do that but even the tuya app seems to have no ability to do that anyway).
 
After manually turning it on (power cycling it then hitting the button to make it boil), then it starts returning DP 1 as the switch:

```
paulr@graviton:~/tmp/tuya$ python test.py
set_status() result {u'devId': u'xxxxxxx', u'dps': {u'1': False, u'5': 95}}
detect_available_dps() result {u'1': False, u'5': 95}
```
If i just leave the kettle alone, at some random point, DP 1 will disappear again for no apparent reason.

Even weirder, if I reboot it, it goes back to only showing one DP and if I then tell it to switch on from python, the DP suddenly turns up again:

```
paulr@graviton:~/tmp/tuya$ python test.py
set_status() result {u'devId': u'xxxxxxx', u'dps': {u'5': 98}}
detect_available_dps() result {u'5': 98}
turn_on()
set_status() result {u'devId': u'xxxxxxx', u'dps': {u'1': True, u'5': 98}}
detect_available_dps() result {u'1': True, u'5': 98}
```

I've since figured out it doesn't really matter if it reports on the existence of DP 1 or not, sending true for DP 1 turns the kettle on.

Not an issue or anything like that, just quite odd.

**Comment by jasonacox** [comment]
Hi @takigama !  Thanks for the note.  This seems like a flaw in the firmware on the kettle.  I have seen some Tuya devices not respond with all of their DPS values but haven't found any that didn't report power DPS if that is controllable.  I suspect if you set up an async monitoring loop like the one below, the device would send "changes" and an event would show up when you power on/off the kettle.  

```python
import tinytuya

d = tinytuya.OutletDevice('DEVICEID', 'DEVICEIP', 'DEVICEKEY')
d.set_version(3.3)
d.set_socketPersistent(True)

print(" > Send Request for Status < ")
payload = d.generate_payload(tinytuya.DP_QUERY)
d.send(payload)

print(" > Begin Monitor Loop <")
while(True):
    # See if any data is available
    data = d.receive()
    print('Received Payload: %r' % data)

    # Send keyalive heartbeat
    print(" > Send Heartbeat Ping < ")
    payload = d.generate_payload(tinytuya.HEART_BEAT)
    d.send(payload)
```

**Reply by takigama** [reply]
Odd output, not entirely sure what I was expecting though to be honest:

```
Received Payload: None
 > Send Heartbeat Ping <
Received Payload: None
 > Send Heartbeat Ping <
Received Payload: None
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'5': 89}, u't': 1652290482}
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'5': 88}, u't': 1652290482}
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'5': 89}, u't': 1652290483}
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'1': True}, u't': 1652290488}
 > Send Heartbeat Ping <
Received Payload: None
 > Send Heartbeat Ping <
Received Payload: None
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'5': 88}, u't': 1652290499}
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'5': 89}, u't': 1652290501}
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'5': 90}, u't': 1652290502}
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'5': 91}, u't': 1652290504}
 > Send Heartbeat Ping <
Received Payload: {u'devId': u'8717320184f3ebedb5ea', u'dps': {u'5': 92}, u't': 1652290506}
```

**Reply by jasonacox** [reply]
Keep in mind that is data that the device is sending without you asking for it.  It shows that any change in temp will result in a state change.  It sent one with power "True" at one point.  Did you try toggling the power while running this?  I'm sure it would send those update to "False" when it was off.  In any case, it is clear that it doesn't value sending DPS 1 as a regular update which explains why you don't always get that in the status request.  That's annoying but likely the way it works.  

The only other thing you could try to force it to tell you the state of DPS 1 is to send the update command:

```python
import tinytuya

d=tinytuya.OutletDevice(DEVICEID, DEVICEIP, DEVICEKEY)
d.set_version(3.3)

# Option for Power Monitoring Smart Plugs - Some require UPDATEDPS to update power data points
payload = d.generate_payload(tinytuya.UPDATEDPS,['1'])
d.send(payload)

# Get the status of the device 
data = d.get_status()
print(data)
```
